File: apps/macos/Sources/Flow/Support/Confetti.swift
import SwiftUI

// Confetti for a 🎉 (#524) — the native half of #514, and a port of the web
// client's `lib/confetti.ts`: when someone celebrates a message, the moment
// should look like a celebration for everyone with the channel open.
//
// Hand-rolled on one `Canvas` per window rather than a particle dependency:
// the whole effect is a second of falling rectangles. Particle positions are a
// pure function of the burst's age, so nothing mutates per frame — a
// `TimelineView(.animation)` redraws, and the overlay disappears entirely once
// the last burst has died, leaving a quiet workspace with no overlay at all.
//
// This file compiles into both apps (iOS pulls `Support/` in via project.yml),
// so the count-went-up rule exists once and neither client can drift from it.

/// Reactions that read as a celebration and so earn a burst. Mirrors
/// `CELEBRATION_EMOJI` in `packages/web/src/lib/confetti.ts`.
enum Celebration {
    static let emoji: Set<String> = ["🎉", "🎊"]
}

/// Celebration counts as we last saw them, per message. Process-level (like
/// `MessageScrollMemory`, and like the web module's `lastSeen` map) so a row
/// that remounts — switching channels and back, a thread opening — doesn't
/// replay bursts for reactions it already showed.
@MainActor
enum CelebrationMemory {
    /// Bound the memory on a long-lived session; oldest messages fall out first.
    static let limit = 2000

    private static var lastSeen: [String: [String: Int]] = [:]
    /// Swift dictionaries are unordered, so insertion order — which is what
    /// makes "oldest first" mean anything — is tracked alongside.
    private static var order: [String] = []

    /// Which celebration emoji were just *added* to this message, folding the
    /// new counts into memory as it goes.
    ///
    /// The first sighting of a message never fires: that is channel load,
    /// scrollback and every re-render of history, where existing reactions are
    /// just state. Only a count that goes up on a message we already rendered
    /// is someone celebrating right now — a local tap and another user's
    /// reaction arriving over the WS both land here the same way. A count that
    /// drops (removal) is recorded silently.
    static func celebrationsAdded(messageId: String, reactions: [ReactionAgg]) -> [String] {
        let before = lastSeen[messageId]
        var now: [String: Int] = [:]
        var added: [String] = []
        for r in reactions where Celebration.emoji.contains(r.emoji) {
            now[r.emoji] = r.count
            if let before, r.count > (before[r.emoji] ?? 0) { added.append(r.emoji) }
        }
        if before == nil {
            if lastSeen.count >= limit, !order.isEmpty {
                lastSeen[order.removeFirst()] = nil
            }
            order.append(messageId)
        }
        lastSeen[messageId] = now
        return added
    }

    /// Forget every message's counts. Tests only.
    static func reset() {
        lastSeen = [:]
        order = []
    }
}

// MARK: - The particle overlay

/// One particle's fixed properties. Its position at any moment is derived from
/// these and the burst's age, so a frame is a draw and never a mutation.
private struct ConfettiSeed {
    var vx: Double
    var vy: Double
    var rot: Double
    var vrot: Double
    var w: Double
    var h: Double
    var color: Color

    static func random() -> ConfettiSeed {
        // Up and outward, with enough spread that no two bursts look alike.
        let angle = -Double.pi / 2 + (Double.random(in: 0...1) - 0.5) * 1.6
        let speed = 180 + Double.random(in: 0...220)
        return ConfettiSeed(
            vx: cos(angle) * speed,
            vy: sin(angle) * speed,
            rot: Double.random(in: 0...Double.pi),
            vrot: (Double.random(in: 0...1) - 0.5) * 12,
            w: 5 + Double.random(in: 0...4),
            h: 8 + Double.random(in: 0...5),
            color: Color(hex: ConfettiPhysics.colors.randomElement() ?? 0xF9C7_4F)
        )
    }
}

/// The numbers, all lifted from the web module so the two look alike.
enum ConfettiPhysics {
    static let colors: [UInt32] = [
        0xF941_43, 0xF372_2C, 0xF9C7_4F, 0x43AA_8B, 0x4D90_8E, 0x9D4E_DD, 0xFF70_A6,
    ]
    /// Particles per burst, and the ceiling across all live bursts — a pile-on
    /// spends what is left of the budget and then stops spawning, so twenty
    /// rapid reactions cost the same frame as three.
    static let burstParticles = 26
    static let maxParticles = 200
    static let life: Double = 1.0
    /// pt/s² — tuned so a particle falls a few dozen points over its short life.
    static let gravity: Double = 900
}

private struct ConfettiBurst: Identifiable {
    let id = UUID()
    let origin: CGPoint
    let startedAt: Date
    let seeds: [ConfettiSeed]
}

/// The live bursts for one window. One controller per `confettiHost()`, handed
/// down the environment — so on macOS, where every window mounts its own root,
/// a burst renders in the window whose pill was reacted to and nowhere else.
@MainActor
final class ConfettiController: ObservableObject {
    /// Redrawing is driven by `TimelineView`; this publishes only when a burst
    /// starts or is reaped, which is what mounts and unmounts the overlay.
    @Published fileprivate private(set) var bursts: [ConfettiBurst] = []

    /// Where each reaction pill currently sits, in the host's coordinate space.
    /// Deliberately *not* published: pills report their frame on every scroll
    /// tick, and a published write there would re-render the transcript.
    private var pillCenters: [String: CGPoint] = [:]

    private static func pillKey(_ messageId: String, _ emoji: String) -> String {
        "\(messageId)\u{1}\(emoji)"
    }

    func notePill(messageId: String, emoji: String, center: CGPoint) {
        pillCenters[Self.pillKey(messageId, emoji)] = center
    }

    func forgetPill(messageId: String, emoji: String) {
        pillCenters[Self.pillKey(messageId, emoji)] = nil
    }

    func pillCenter(messageId: String, emoji: String) -> CGPoint? {
        pillCenters[Self.pillKey(messageId, emoji)]
    }

    /// Particles currently in flight, across every live burst.
    var liveParticleCount: Int { bursts.reduce(0) { $0 + $1.seeds.count } }

    /// Throw a short burst from a point in the host's coordinate space.
    /// Best-effort and decorative throughout: out of budget means no confetti,
    /// and the reaction itself is unaffected either way.
    func burst(at point: CGPoint, now: Date = Date()) {
        let room = ConfettiPhysics.maxParticles - liveParticleCount
        guard room > 0 else { return }
        let count = min(ConfettiPhysics.burstParticles, room)
        let burst = ConfettiBurst(
            origin: point,
            startedAt: now,
            seeds: (0..<count).map { _ in ConfettiSeed.random() }
        )
        bursts.append(burst)
        // Reaped on a timer rather than mid-draw: a `Canvas` closure must stay
        // a pure render, and this is the one place the array shrinks.
        Task { [weak self] in
            try? await Task.sleep(for: .milliseconds(Int(ConfettiPhysics.life * 1000) + 100))
            self?.bursts.removeAll { $0.id == burst.id }
        }
    }
}

/// The overlay itself: pinned over the whole host, click-through, invisible to
/// assistive tech, and absent entirely when nothing is flying.
private struct ConfettiCanvas: View {
    @ObservedObject var controller: ConfettiController

    var body: some View {
        if controller.bursts.isEmpty {
            Color.clear.frame(width: 0, height: 0)
        } else {
            TimelineView(.animation) { timeline in
                Canvas { ctx, _ in
                    for burst in controller.bursts {
                        draw(burst, at: timeline.date, in: &ctx)
                    }
                }
            }
            .allowsHitTesting(false)
            .accessibilityHidden(true)
        }
    }

    private func draw(_ burst: ConfettiBurst, at date: Date, in ctx: inout GraphicsContext) {
        let age = date.timeIntervalSince(burst.startedAt)
        guard age >= 0, age < ConfettiPhysics.life else { return }
        // Fades over the last third of its life rather than blinking out.
        ctx.opacity = min(1, (ConfettiPhysics.life - age) / (ConfettiPhysics.life * 0.35))
        for seed in burst.seeds {
            let x = burst.origin.x + seed.vx * age
            let y = burst.origin.y + seed.vy * age + 0.5 * ConfettiPhysics.gravity * age * age
            let rect = CGRect(x: -seed.w / 2, y: -seed.h / 2, width: seed.w, height: seed.h)
            let transform = CGAffineTransform(translationX: x, y: y)
                .rotated(by: seed.rot + seed.vrot * age)
            ctx.fill(Path(rect).applying(transform), with: .color(seed.color))
        }
    }
}

// MARK: - Wiring

/// The coordinate space pills report into and the canvas draws in.
private let confettiSpace = "flow.confetti"

private struct ConfettiControllerKey: EnvironmentKey {
    static let defaultValue: ConfettiController? = nil
}

extension EnvironmentValues {
    /// nil outside a `confettiHost()` — every source below no-ops rather than
    /// trapping, because decoration must never be able to break a reaction.
    fileprivate var confettiController: ConfettiController? {
        get { self[ConfettiControllerKey.self] }
        set { self[ConfettiControllerKey.self] = newValue }
    }
}

private struct ConfettiHostModifier: ViewModifier {
    @StateObject private var controller = ConfettiController()

    func body(content: Content) -> some View {
        content
            .coordinateSpace(name: confettiSpace)
            .environment(\.confettiController, controller)
            .overlay(ConfettiCanvas(controller: controller))
    }
}

/// Reports a reaction pill's position so a burst can start where the pill is.
/// Cheap by construction: the frame lands in unpublished storage, so tracking
/// it through a scroll costs no re-render of the transcript.
private struct ConfettiPillModifier: ViewModifier {
    let messageId: String
    let emoji: String
    @Environment(\.confettiController) private var controller

    func body(content: Content) -> some View {
        content.background(
            GeometryReader { geo in
                let frame = geo.frame(in: .named(confettiSpace))
                Color.clear
                    .onAppear { note(frame) }
                    .onChange(of: frame) { _, new in note(new) }
                    .onDisappear { controller?.forgetPill(messageId: messageId, emoji: emoji) }
            }
        )
    }

    private func note(_ frame: CGRect) {
        controller?.notePill(
            messageId: messageId, emoji: emoji,
            center: CGPoint(x: frame.midX, y: frame.midY)
        )
    }
}

/// Watches a message's reaction aggregates and bursts from the pill that just
/// went up. Goes on the *message row*, not on the reaction row: a message with
/// no reactions yet renders no reaction row at all, and if the memory only
/// started there, the first 🎉 on a message would look like a first sighting
/// and stay silent — which is the one case anyone tries first.
private struct CelebrationBurstsModifier: ViewModifier {
    let messageId: String
    let reactions: [ReactionAgg]
    @Environment(\.confettiController) private var controller
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    func body(content: Content) -> some View {
        content
            // `initial: true` is what records the first sighting — and a first
            // sighting is silent by construction, so channel load, scrollback
            // and a row remounting animate nothing.
            .onChange(of: reactions, initial: true) { _, current in
                fire(current)
            }
    }

    private func fire(_ current: [ReactionAgg]) {
        let added = CelebrationMemory.celebrationsAdded(messageId: messageId, reactions: current)
        // Memory is folded in even under Reduce Motion, so turning it off
        // mid-session doesn't then replay a backlog of old reactions.
        guard !added.isEmpty, !reduceMotion, let controller else { return }
        // One hop, so a pill that only just appeared has been laid out and can
        // report where it is before we ask.
        Task { @MainActor in
            for emoji in added {
                guard let origin = controller.pillCenter(messageId: messageId, emoji: emoji)
                    ?? controller.pillCenter(messageId: messageId, emoji: "")
                else { continue }
                controller.burst(at: origin)
            }
        }
    }
}

extension View {
    /// Mount the confetti overlay and the coordinate space its sources report
    /// into. Once, at the root of a window.
    func confettiHost() -> some View {
        modifier(ConfettiHostModifier())
    }

    /// One reaction pill: where a burst for this emoji should start.
    func confettiPill(messageId: String, emoji: String) -> some View {
        modifier(ConfettiPillModifier(messageId: messageId, emoji: emoji))
    }

    /// The reaction row as a whole — the fallback origin for an emoji whose
    /// pill has only just appeared and has not reported a frame yet. The empty
    /// emoji is the key, which no real pill can claim.
    func confettiReactionRow(messageId: String) -> some View {
        modifier(ConfettiPillModifier(messageId: messageId, emoji: ""))
    }

    /// The message row: fires a burst when a 🎉/🎊 count goes up on it.
    func celebrationBursts(messageId: String, reactions: [ReactionAgg]) -> some View {
        modifier(CelebrationBurstsModifier(messageId: messageId, reactions: reactions))
    }
}

File: packages/web/src/lib/confetti.ts
// Confetti for a 🎉 (#514) — when someone celebrates a message, the moment
// should look like a celebration for everyone with the channel open.
//
// Hand-rolled on one shared canvas rather than a particle library: the whole
// effect is a second of falling rectangles, and a dependency for that would
// cost more to keep than to write. The canvas is created on the first burst
// and removed again when the last particle dies, so a quiet workspace carries
// no overlay at all.

/** Reactions that read as a celebration and so earn a burst. */
export const CELEBRATION_EMOJI = ['🎉', '🎊'];

/** Just the shape of a reaction this module cares about. */
export interface ReactionCount {
  emoji: string;
  count: number;
}

/**
 * Celebration counts as this module last saw them, per message. Module-level
 * (like `scrollMemory` in MessageList) so a row that remounts — switching
 * channels and back, a thread opening — doesn't replay bursts for reactions it
 * already showed.
 */
const lastSeen = new Map<string, Map<string, number>>();
/** Bound the memory on a long-lived tab; oldest messages fall out first. */
const SEEN_LIMIT = 2000;

/**
 * Which celebration emoji were just *added* to this message, folding the new
 * counts into memory as it goes.
 *
 * The first sighting of a message never fires: that is channel load, scrollback
 * and every re-render of history, where existing reactions are just state.
 * Only a count that goes up on a message we already rendered is someone
 * celebrating right now — a local click or another user's reaction arriving
 * over the WS both land here the same way. A count that drops (removal) is
 * recorded silently.
 */
export function celebrationsAdded(messageId: string, reactions: ReactionCount[]): string[] {
  const before = lastSeen.get(messageId);
  const now = new Map<string, number>();
  const added: string[] = [];
  for (const r of reactions) {
    if (!CELEBRATION_EMOJI.includes(r.emoji)) continue;
    now.set(r.emoji, r.count);
    if (before && r.count > (before.get(r.emoji) ?? 0)) added.push(r.emoji);
  }
  if (!before && lastSeen.size >= SEEN_LIMIT) {
    const oldest = lastSeen.keys().next();
    if (!oldest.done) lastSeen.delete(oldest.value);
  }
  lastSeen.set(messageId, now);
  return added;
}

/** Forget every message's counts. Tests only. */
export function resetCelebrationMemory(): void {
  lastSeen.clear();
}

/** Nobody who asked the OS to stop moving things gets confetti. */
function prefersReducedMotion(): boolean {
  return window.matchMedia?.('(prefers-reduced-motion: reduce)').matches === true;
}

const COLORS: [string, ...string[]] = ['#f94144', '#f3722c', '#f9c74f', '#43aa8b', '#4d908e', '#9d4edd', '#ff70a6'];
/** Particles per burst, and the ceiling across all live bursts — a pile-on
 * spends what is left of the budget and then stops spawning, so twenty rapid
 * reactions cost the same frame as three. */
const BURST_PARTICLES = 26;
const MAX_PARTICLES = 200;
const LIFE_MS = 1000;
/** px/s² — tuned so a particle falls a few dozen pixels over its short life. */
const GRAVITY = 900;

interface Particle {
  x: number;
  y: number;
  vx: number;
  vy: number;
  rot: number;
  vrot: number;
  w: number;
  h: number;
  color: string;
  age: number;
}

let canvas: HTMLCanvasElement | null = null;
let ctx: CanvasRenderingContext2D | null = null;
let particles: Particle[] = [];
let frame = 0;
let lastFrameAt = 0;

/** The overlay: viewport-fixed, click-through, above everything, invisible to
 * assistive tech. It never affects layout, so it can live on document.body. */
function ensureCanvas(): CanvasRenderingContext2D | null {
  if (ctx) return ctx;
  const el = document.createElement('canvas');
  el.setAttribute('data-testid', 'confetti-canvas');
  el.setAttribute('aria-hidden', 'true');
  el.style.cssText = 'position:fixed;inset:0;width:100%;height:100%;pointer-events:none;z-index:9999';
  const c2d = el.getContext?.('2d') ?? null;
  if (!c2d) return null;
  document.body.appendChild(el);
  canvas = el;
  ctx = c2d;
  return ctx;
}

function teardown(): void {
  if (frame) cancelAnimationFrame(frame);
  frame = 0;
  particles = [];
  canvas?.remove();
  canvas = null;
  ctx = null;
}

function resize(): void {
  if (!canvas || !ctx) return;
  const dpr = window.devicePixelRatio || 1;
  const w = Math.round(window.innerWidth * dpr);
  const h = Math.round(window.innerHeight * dpr);
  if (canvas.width === w && canvas.height === h) return;
  canvas.width = w;
  canvas.height = h;
  ctx.setTransform(dpr, 0, 0, dpr, 0, 0);
}

function step(now: number): void {
  frame = 0;
  if (!ctx || !canvas) return;
  // Clamp dt so a backgrounded tab doesn't resume with one enormous jump, and
  // never let it go negative — a clock that runs backwards ages particles
  // backwards, and they stop dying.
  const dt = Math.max(0, Math.min(now - lastFrameAt, 50)) / 1000;
  lastFrameAt = now;
  resize();
  ctx.clearRect(0, 0, window.innerWidth, window.innerHeight);
  const alive: Particle[] = [];
  for (const p of particles) {
    p.age += dt * 1000;
    if (p.age >= LIFE_MS) continue;
    p.vy += GRAVITY * dt;
    p.x += p.vx * dt;
    p.y += p.vy * dt;
    p.rot += p.vrot * dt;
    alive.push(p);
    // Fades over the last third of its life rather than blinking out.
    ctx.globalAlpha = Math.min(1, (LIFE_MS - p.age) / (LIFE_MS * 0.35));
    ctx.save();
    ctx.translate(p.x, p.y);
    ctx.rotate(p.rot);
    ctx.fillStyle = p.color;
    ctx.fillRect(-p.w / 2, -p.h / 2, p.w, p.h);
    ctx.restore();
  }
  ctx.globalAlpha = 1;
  particles = alive;
  if (particles.length === 0) teardown();
  else frame = requestAnimationFrame(step);
}

/**
 * Throw a short burst of confetti from a point in viewport coordinates.
 * Best-effort and decorative throughout: no canvas, no motion budget, no
 * confetti — the reaction itself is unaffected either way.
 */
export function burstConfetti(x: number, y: number): void {
  if (typeof document === 'undefined' || typeof window === 'undefined') return;
  if (prefersReducedMotion()) return;
  const c2d = ensureCanvas();
  if (!c2d) return;
  const room = MAX_PARTICLES - particles.length;
  if (room <= 0) return;
  resize();
  for (let i = 0; i < Math.min(BURST_PARTICLES, room); i++) {
    // Up and outward, with enough spread that no two bursts look alike.
    const angle = -Math.PI / 2 + (Math.random() - 0.5) * 1.6;
    const speed = 180 + Math.random() * 220;
    particles.push({
      x,
      y,
      vx: Math.cos(angle) * speed,
      vy: Math.sin(angle) * speed,
      rot: Math.random() * Math.PI,
      vrot: (Math.random() - 0.5) * 12,
      w: 5 + Math.random() * 4,
      h: 8 + Math.random() * 5,
      color: COLORS[Math.floor(Math.random() * COLORS.length)] ?? COLORS[0],
      age: 0,
    });
  }
  if (!frame) {
    lastFrameAt = performance.now();
    frame = requestAnimationFrame(step);
  }
}
